gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/GradleCacheExplorer.java

```diff
@@ -0,0 +1,175 @@
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.progress.ProcessCanceledException;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GradleCacheExplorer {
+
+    private static final Logger log = LoggerFactory.getLogger(GradleCacheExplorer.class);
+    private static final String GRADLE_CACHE_PATH = System.getProperty("user.home") + "/.gradle/caches/modules-2/";
+    private final Cache<String, Set<String>> cache =
+            Caffeine.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build();
+
+    private final List<String> projectUrls;
+
+    public GradleCacheExplorer(List<String> projectUrls) {
+        this.projectUrls = projectUrls;
+    }
+
+    public final Set<String> getCompletions(DependencyGroup input) {
+        String parsedInput = String.join(".", input.parts());
+        ProgressIndicator indicator = ProgressManager.getInstance().getProgressIndicator();
+        if (indicator == null) {
+            return Collections.emptySet();
+        }
+
+        try {
+            Callable<Set<String>> task = () -> extractStrings(indicator);
+            Future<Set<String>> future = ApplicationManager.getApplication().executeOnPooledThread(task);
+            Set<String> results =
+                    com.intellij.openapi.application.ex.ApplicationUtil.runWithCheckCanceled(future::get, indicator);
+
+            if (parsedInput.isEmpty()) {
+                return results;
+            }
+
+            return results.stream()
+                    .filter(result -> result.startsWith(parsedInput))
+                    .map(result -> result.substring(parsedInput.length() + 1))
+                    .collect(Collectors.toSet());
+        } catch (ProcessCanceledException e) {
+            log.debug("Operation was cancelled", e);
+        } catch (Exception e) {
+            log.warn("Failed to get completions", e);
+        }
+        return Collections.emptySet();
+    }
+
+    private Set<String> extractStrings(ProgressIndicator indicator) {
+        return cache.get("metadata", key -> {
+            try (Stream<Path> metadataFolders = Files.list(Paths.get(GRADLE_CACHE_PATH))
+                    .filter(path -> path.getFileName().toString().startsWith("metadata-"))) {
```

Comment:
nit: can you move this filter outside of the `try-with-resources` initialiser? makes it a bit hard to read